components/camel-spring-batch/src/main/java/org/apache/camel/component/spring/batch/SpringBatchProducer.java

```diff
@@ -44,8 +46,33 @@ public SpringBatchProducer(SpringBatchEndpoint endpoint, JobLauncher jobLauncher
 
     @Override
     public void process(Exchange exchange) throws Exception {
+
         JobParameters jobParameters = prepareJobParameters(exchange.getIn().getHeaders());
-        JobExecution jobExecution = jobLauncher.run(job, jobParameters);
+        String messageJobName = jobParameters.getString(SpringBatchComponent.DYNAMIC_JOBNAME);
+
+        Job job2run = this.job;
+
+        if (messageJobName != null)
+        {
+            Job dynamicJob = CamelContextHelper.mandatoryLookup(getEndpoint().getCamelContext(), messageJobName, Job.class);
+
+            job2run = dynamicJob;
+
+            if (job2run == null)
+            {
+                exchange.setException(new CamelExchangeException("Found header " + SpringBatchComponent.DYNAMIC_JOBNAME +
+                        " with value " +messageJobName + " but could not find a Job in camel context", exchange));
+                return;
+            }
+        }
+
+        if (job2run == null) {
```

Comment:
No it's not redundant. if the jobName is dynamic, then the job will be set to null by the Endpoint.